src/librustc_mir/transform/nll/infer.rs

```diff
@@ -0,0 +1,206 @@
+use super::{Region, RegionIndex};
+use std::mem;
+use rustc::infer::InferCtxt;
+use rustc::mir::{Location, Mir};
+use rustc_data_structures::indexed_vec::{Idx, IndexVec};
+use rustc_data_structures::fx::FxHashSet;
+
+pub struct InferenceContext {
+    definitions: IndexVec<RegionIndex, VarDefinition>,
+    constraints: IndexVec<ConstraintIndex, Constraint>,
+    errors: IndexVec<InferenceErrorIndex, InferenceError>,
+}
+
+pub struct InferenceError {
+    pub constraint_point: Location,
+    pub name: (), // TODO(nashenas88) RegionName
+}
+
+newtype_index!(InferenceErrorIndex);
```

Comment:
Creating a newtype index doesn't seem worth it to me here -- I think making a newtype'd index is only worth it if we are passing around these indices a lot. I don't envision us doing anything here but iterating through the errors, right?